api/v1alpha1/httproute_types.go

```diff
@@ -190,6 +200,75 @@ type HTTPRouteAction struct {
 	Extension *core.TypedLocalObjectReference `json:"extension"`
 }
 
+// TLSConfig defines the schema of a TLS configuration.
+type TLSConfig struct {
+	// Termination defines how to terminate TLS connections.
+	//
+	// If unspecified, TLS termination type "Edge" will be used.
+	//
+	// Support: core
+	//
+	// +optional
+	Termination TLSTerminationType `json:"termination,omitempty"`
+
+	// Certificate is a reference to a Kubernetes object containing
+	// the identity certificate, key and CA certificate used to terminate
+	// the TLS connection. Certificate is used with "Reencrypt" and "Edge"
+	// TLS termination types.
+	//
+	// If apiGroup and kind are empty, will default to a Kubernetes Secret
+	// resources.
+	//
+	// If unspecified, the certificate of the binding gateway will be used.
+	//
+	// Support: Core (Kubernetes Secrets)
+	// Support: Implementation-specific (For other resource types)
+	//
+	// +optional
+	Certificate core.TypedLocalObjectReference `json:"certificate,omitempty"`
+
+	// DestinationCACertificate is a reference to a CA certificate used
+	// for establishing a TLS connection with the final destination when
+	// using TLS termination type "Reencrypt".
+	//
+	// Here is a ConfigMap example (in yaml):
+	//
+	// apiVersion: v1
+	// kind: ConfigMap
+	// metadata:
+	//  name: my-dest-svc-ca
+	//  namespace: my-dest-svc-namespace
+	//  data:
+	//    ca-bundle.crt: |
+	//      -----BEGIN CERTIFICATE-----
+	//      Destination Service CA Certificate Bundle.
+	//      -----END CERTIFICATE-----
+	//
+	// Support: Core (Kubernetes ConfigMap)
+	// Support: Implementation-specific (For other resource types)
+	//
+	// +optional
+	DestinationCACertificate core.TypedLocalObjectReference `json:"destinationCACertificate,omitempty"`
+}
+
+// TLSTerminationType specifies where TLS connections will terminate.
+type TLSTerminationType string
+
+const (
+	// TLSTerminationEdge terminates the TLS connection at the gateway.
+	TLSTerminationEdge TLSTerminationType = "Edge"
+
+	// TLSTerminationPassthrough terminates the TLS connection at the
+	// destination service. The destination service is responsible for
+	// decrypting data from the connection.
+	TLSTerminationPassthrough TLSTerminationType = "Passthrough"
+
+	// TLSTerminationReencrypt terminates the TLS connection at the gateway.
+	// The gateway creates an encrypted connection to the destination service
+	// using the provided certificate from DestinationCACertificate.
+	TLSTerminationReencrypt TLSTerminationType = "Reencrypt"
```

Comment:
I was confused when I saw this option.

1. If we are doing passthrough, it should not be an HTTPRoute, I'd rather see that in TCPRoute or a TLSRoute, since we can't even know the protocol or muck around in it as it's an encrypted stream.
2. Upstream connection details should not be part of the route level details but `HTTPRouteAction` maybe. Each service or action can potentially tweak this part a little differently.

Reply:
Right. I think that it is useful to think about this hop-by-hop. What does the hop from client to gateway look like? What does the hop from gateway to the next endpoint look like? I this model the name "reencrypt" doesn't really apply because we are focused only on hops.